src/sentry/monitors/tasks.py

```diff
@@ -428,40 +435,102 @@ def detect_broken_monitor_envs():
         monitor__status=ObjectStatus.ACTIVE,
         monitor__is_muted=False,
         monitor_environment__is_muted=False,
-        # TODO(davidenwang): When we want to email users, remove this filter
-        monitorenvbrokendetection__isnull=True,
+        # TODO(davidenwang): Once we start disabling environments, change accordingly
+        monitorenvbrokendetection__user_notified_timestamp=None,
     )
-    for open_incident in RangeQuerySetWrapper(
-        open_incidents_qs,
-        order_by="starting_timestamp",
-        step=1000,
-    ):
+    org_ids_with_open_incidents = (
+        open_incidents_qs.all().values_list("monitor__organization_id", flat=True).distinct()
+    )
+
+    for org_id in org_ids_with_open_incidents:
         try:
-            organization = Organization.objects.get_from_cache(
-                id=open_incident.monitor.organization_id
-            )
+            organization = Organization.objects.get_from_cache(id=org_id)
             if not features.has(
                 "organizations:crons-broken-monitor-detection", organization=organization
             ):
                 continue
         except Organization.DoesNotExist:
             continue
 
-        # verify that the most recent check-ins have been failing
-        recent_checkins = (
-            MonitorCheckIn.objects.filter(monitor_environment=open_incident.monitor_environment)
-            .order_by("-date_added")
-            .values("status")[:NUM_CONSECUTIVE_BROKEN_CHECKINS]
+        # Map user email to a dictionary of monitors and their earliest incident start date amongst its broken environments
+        user_broken_envs = defaultdict(
+            lambda: defaultdict(
+                lambda: {"environment_names": [], "earliest_start": django_timezone.now()}
+            )
         )
-        if len(recent_checkins) != NUM_CONSECUTIVE_BROKEN_CHECKINS or not all(
-            checkin["status"] in [CheckInStatus.ERROR, CheckInStatus.TIMEOUT, CheckInStatus.MISSED]
-            for checkin in recent_checkins
+        orgs_open_incidents = (
+            open_incidents_qs.all()
+            .select_related("monitor_environment")
+            .filter(monitor__organization_id=org_id)
+        )
+        # Query for all the broken incidents within the current org we are processing
+        for open_incident in RangeQuerySetWrapper(
```

Comment:
not sure if I need this RangeQuerySetWrapper here, each org could technically have 1500 monitors each with 1000 environments, so figure it would just be safe to do this